src/main/cpp/subsystems/DriveSubsystem.cpp

```diff
@@ -65,131 +66,63 @@ DriveSubsystem::DriveSubsystem(Vision *vision)
-
-    if (angleDif < 0.45 * std::numbers::pi) {
-      m_currentTranslationDir = SwerveUtils::StepTowardsCircular(
-          m_currentTranslationDir, inputTranslationDir,
-          directionSlewRate * elapsedTime);
-      m_currentTranslationMag = m_magLimiter.Calculate(inputTranslationMag);
-    } else if (angleDif > 0.85 * std::numbers::pi) {
-      if (m_currentTranslationMag >
-          1e-4) {  // some small number to avoid floating-point errors with
-                   // equality checking
-        // keep currentTranslationDir unchanged
-        m_currentTranslationMag = m_magLimiter.Calculate(0.0);
-      } else {
-        m_currentTranslationDir =
-            SwerveUtils::WrapAngle(m_currentTranslationDir + std::numbers::pi);
-        m_currentTranslationMag = m_magLimiter.Calculate(inputTranslationMag);
-      }
-    } else {
-      m_currentTranslationDir = SwerveUtils::StepTowardsCircular(
-          m_currentTranslationDir, inputTranslationDir,
-          directionSlewRate * elapsedTime);
-      m_currentTranslationMag = m_magLimiter.Calculate(0.0);
-    }
-    m_prevTime = currentTime;
-
-    xSpeedCommanded = m_currentTranslationMag * cos(m_currentTranslationDir);
-    ySpeedCommanded = m_currentTranslationMag * sin(m_currentTranslationDir);
-    m_currentRotation = m_rotLimiter.Calculate(rot.value());
-
-  } else {
-    xSpeedCommanded = xSpeed.value();
-    ySpeedCommanded = ySpeed.value();
-    m_currentRotation = rot.value();
-  }
-
-  // Convert the commanded speeds into the correct units for the drivetrain
-  units::meters_per_second_t xSpeedDelivered =
-      xSpeedCommanded * DriveConstants::kMaxSpeed;
-  units::meters_per_second_t ySpeedDelivered =
-      ySpeedCommanded * DriveConstants::kMaxSpeed;
-  units::radians_per_second_t rotDelivered =
-      m_currentRotation * DriveConstants::kMaxAngularSpeed;
-
-  auto states = kDriveKinematics.ToSwerveModuleStates(
-      fieldRelative
-          ? frc::ChassisSpeeds::Discretize(frc::ChassisSpeeds::FromFieldRelativeSpeeds(
-                xSpeedDelivered, ySpeedDelivered, rotDelivered,
-                frc::Rotation2d(units::degree_t{-m_gyro.GetAngle()})), DriveConstants::kLoopTime)
-          : frc::ChassisSpeeds{xSpeedDelivered, ySpeedDelivered, rotDelivered});
-
-  kDriveKinematics.DesaturateWheelSpeeds(&states, DriveConstants::kMaxSpeed);
+
+  auto now = frc::Timer::GetFPGATimestamp();
+  auto dif = now - driveLoopTime;
+
+  auto states =
+      m_driveKinematics.ToSwerveModuleStates(frc::ChassisSpeeds::Discretize(
+          fieldRelative ? frc::ChassisSpeeds::FromFieldRelativeSpeeds(
+                              xSpeed * DriveConstants::kMaxSpeed.value(),
+                              ySpeed * DriveConstants::kMaxSpeed.value(),
+                              rot * DriveConstants::kMaxAngularSpeed.value(),
+                              m_gyro.GetRotation2d())
+                        : frc::ChassisSpeeds{xSpeed, ySpeed, rot},
+          dif));
```

Comment:
make sure we aren't accidentally reintroducing the bug from emcc

Reply:
Mind expanding more on that bug?

Reply:
After more testing, I don't think we have a bug here.

Reply:
The bug is related to the dif not being initialized with a value, if it’s zero it’ll affect drive in the first loop

Reply:
If it’s initialized with a value we should be safe